### ISSUE TITLE
Edge linestyles not always working

### ISSUE THREAD

**Issue by lmshk** [MakieOrg/GraphMakie.jl#93]
tl;dr: Using `edge_attr = (; linestyle = [...])` requires setting `edge_plottype = :beziersegments` too.

I was trying to use the `linestyle` edge attribute to represent edge types, but passing a `Vector{Symbol}` of line styles per edge sometimes caused plotting to fail:

```
ERROR: LoadError: MethodError: no method matching *(::Symbol, ::Float64)
[...]
Stacktrace:
  [1] _broadcast_getindex_evalf
    @ ./broadcast.jl:670 [inlined]
  [2] _broadcast_getindex
    @ ./broadcast.jl:643 [inlined]
  [3] getindex
    @ ./broadcast.jl:597 [inlined]
  [4] copy
    @ ./broadcast.jl:899 [inlined]
  [5] materialize(bc::Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{1}, Nothing, typeof(*), Tuple{Vector{Symbol}, Float64}})
    @ Base.Broadcast ./broadcast.jl:860
  [6] (::GLMakie.var"#189#190"{GLMakie.Screen{GLFW.Window}, Makie.Scene, MakieCore.LineSegments{Tuple{Vector{GeometryBasics.Point{2, Float32}}}}})(gl_attributes::Dict{Symbol, Any})
    @ GLMakie ~/.julia/packages/GLMakie/x68HK/src/drawing_primitives.jl:278
[...]
```

After investigating, I noticed that this only happens if there are no double- or self-edges in the graph. The reason is that in these cases (or if the graph is too large), `edge_plottype = Makie.automatic` makes GraphMakie choose `Makie.linesegments!` to plot the edges, which interprets the `Vector` as defining a single `linestyle` pattern to be used for all the edges. (That behavior is undocumented for `Makie.linesegments`, though the `Makie.lines` documentation has an example.) The GraphMakie `beziersegments!` recipe, on the other hand, works as expected because [it lifts the properties itself](https://github.com/MakieOrg/GraphMakie.jl/blob/061addf41458ea676cdf1e956df90932389f8b2b/src/recipes.jl#L589). Therefore, a workaround (for smaller graphs) is to explicitly set `edge_plottype = :beziersegments`.

Example:

```julia
using GraphMakie, GLMakie, Graphs

graphplot(
    DiGraph([Edge(1 => 2), Edge(2 => 1)]),
    edge_attr = (; linestyle = [:dot, :dash]),
)
# works

graphplot(
    DiGraph([Edge(1 => 2), Edge(2 => 3)]),
    edge_attr = (; linestyle = [:dot, :dash]),
)
# MethodError: no method matching *(::Symbol, ::Float64)

graphplot(
    DiGraph([Edge(1 => 2), Edge(2 => 3)]),
    edge_attr = (; linestyle = [:dot, :dash]),
    edge_plottype = :beziersegments,
)
# works
```

Conversely, using the `Float64[]` linestyle variant (defining a custom pattern) works in the `:linesegments` case, while for `:beziersegments` it depends on the number of nodes in the graph: if the pattern length is equal to `Graphs.nv`, the error is "Pattern needs to be a Vector of floats. Found: Float64", otherwise it seems to work.

This is not strictly a bug, because nothing promises that this should work (so feel free to close this), but I wanted to file this anyway because the behavior surprised me and it took me a while to understand what's going on.

My preferred solution would be [`Makie.linesegments` accepting a vector of linestyles](https://github.com/MakieOrg/Makie.jl/issues/1598), but otherwise maybe GraphMakie should plot individual `Makie.lines` and lift the properties like in the bezier case to ensure consistency?

**Comment by hexaeder** [comment]
You think the proposed solutions are sufficient?

I don't like auto promoting to `beziersegments` because of performance reasons. I think it is a bit different for curvy edges, because there both options (beziersegments and linesegments) have a clear usecase. I figured in this case it would be better to just provide a better error message which tells the user to specify the plottype.

**Comment by lmshk** [comment]
I think that the specific graph being plotted should not influence whether this gets thrown for otherwise equal properties. More generally, that is probably the case for other vector-valued per-item properties too if there are any that Makie does not itself support. (?)

I'm not sure what makes the bezier case slow, is it the interpolation or is it the fact that every edge then has its own `lines`? Because if it's only the former, maybe the right solution would be to do `Graphs.ne` many `lines` plots instead of a single `linesegments`, and lift all properties in the same way that `beziersegments` uses? Or even to make `beziersegments` use zero interpolated points if the curvature is zero, subsuming the whole `linearsegments` case (and explicitly forcing zero curvature should it be required to straighten the lines), thus eliminating the distinction? Or am I missing something?

Otherwise if the performance issues come from using many `lines`, one per edge, I think that the better error message should get thrown if `linestyle` (or any property that Makie does not support per-item vector definitions for) is a `Vector` and `edge_plottype` is not `:beziersegments`. In the proposed solution, even with the better error, it is not obvious for somebody setting the `linestyle` vector that this issue might occur because they may not have tried with a graph that would trigger it.

**Comment by hexaeder** [comment]
> Or even to make beziersegments use zero interpolated points if the curvature is zero

Thats what happens in the moment. `BezierPath <: AbstractPath` can be an arbitrary path, that one gets discretized. If there is no curvature, the path for the edge is a `Line <: AbstractPath` with trivial discretization of just start end endpoint. In mixed graphs (some `Line`s some `BezierPath`s) this means the edge pathes array gets typeunstable with dynamic dispatch (at some point someone should investigate whether the type instability is worth the elided discretization). For just `Lines` it should be relatively fast. However if you plot graphs with a few thousand nodes it used to be quite slow, presumably because of all the individual plots. Thats why I don't want to give up the `linesegments` path which should be fine for most of the plotted graphs. Even though it would be much nicer to not have that distinction.

I am not aware of any other properties where this happens in the moment. So if I understand you correctly you'd prefer the combination of `edge_plottype = automatic => becomes beziersegments` to fail in combination with a vector of linestyles, forcing the user to always set `edge_plottype` explicitly? I'm fine with taht.

**Comment by lmshk** [comment]
> In mixed graphs (some Lines some BezierPaths) this means the edge pathes array gets typeunstable with dynamic dispatch

> it used to be quite slow, presumably because of all the individual plots

Are you saying the slowdown is mostly due to the former, the latter, or both?

> So if I understand you correctly you'd prefer the combination of edge_plottype = automatic => becomes beziersegments to fail in combination with a vector of linestyles, forcing the user to always set edge_plottype explicitly?

Yes, I'd like any use of per-edge (vector) `linestyle` to fail with the new error unless `edge_plottype = :beziersegments` is set explicitly, to prevent the surprise of this error occurring inconsistently only for some graphs. (This would solve the issue for me.)

**Comment by hexaeder** [comment]
> Are you saying the slowdown is mostly due to the former, the latter, or both?

In mixed graphs (curvy and straight lines) probably both. In graphs without curvy lines where everything is just a line it is slow because of individual plot objects per edge.
